test/support/utils.rb

```diff
@@ -73,18 +74,32 @@ def create_pseudo_terminal(boot_options: "-r debug/run")
               when %r{Really quit\? \[Y/n\]}
                 cmd = @queue.pop
                 write.puts(cmd)
+              when /REPL ERROR/
+                error_messages = [line]
+
+                while line = read.gets
+                  case line.chomp
+                  # it means all error messages have been printed out
+                  when REPL_PROMPT
+                    break
+                  else
+                    error_messages << line
+                  end
+                end
+
+                raise "Debugger terminated because of: #{error_messages.join}"
               end
+
               @backlog.push(line)
               @last_backlog.push(line)
             end
+
+            assert_empty_queue
           end
         rescue Errno::EIO => e
-          if @queue.empty?
-            # result of `gets` return this exception in some platform
-            # https://github.com/ruby/ruby/blob/master/ext/pty/pty.c#L729-L736
-          else
-            assert false, create_message(e.message)
-          end
+          # result of `gets` return this exception in some platform
+          # https://github.com/ruby/ruby/blob/master/ext/pty/pty.c#L729-L736
+          assert_empty_queue
```

Comment:
If you define this method, `e.message` won't show in error message.

Reply:
it's a platform-specific implementation detail to say "the terminal is finished". the user only needs know that the terminal is finished while there still are commands in the queue.

Reply:
Can we fix such as https://github.com/ruby/debug/pull/87#discussion_r651448313 and https://github.com/ruby/debug/pull/87#discussion_r651448425 ?
They are just tips. You can change those suggestions.

Reply:
Oh, I couldn't check your reply. Sorry for confused you.

Reply:
> it's a platform-specific implementation detail to say "the terminal is finished". the user only needs know that the terminal is finished while there still are commands in the queue.

In this stage, we have to improve test framework. This change makes it a little harder to understand cause of the error while developing test framework. 
What do you think?

Reply:
the original implementation only works on certain platforms. on macOS, it doesn't raise the error and the tests just pass.

so I added this PR to specifically catch `REPL ERROR` message and raises it, which is a more consistent way to detect debugger's exceptions.

Reply:
of course I may have missed some cases. do you have an example on what errors would become harder to undertand after this PR?

Reply:
There seems to be miscommunication between us.

My point is It's better not to change https://github.com/ruby/debug/pull/87/files/662d653fd1bb478a81a42a53f2dcac7f39c6de5c#diff-ee14bf65f14927f7778a246f90c5cec0c581be3e2658c2a8a3726c59db5be3f9L82-L87.

In this stage, we have to improve test framework. This change makes it a little harder to understand cause of the error while developing test framework.

> so I added this PR to specifically catch REPL ERROR message and raises it, which is a more consistent way to detect debugger's exceptions.

I agree with your opinion.

Reply:
> In this stage, we have to improve test framework. This change makes it a little harder to understand cause of the error while developing test framework.

sorry, I still don't get this part. 
if a part of the test framework (like an assertion method) has issue, it won't enter this block, right?
if the problem is in the extra patch, they should now be caught because of `REPL ERROR`.

can you provide me a concrete example on what kind of error you have in mind? thanks

Reply:
@ono-max hey if you don't have time to finish this discussion now, can we move it to another issue? as I've said, the current test framework doesn't work for mac users and I can't trust the test result on my machine, which really bothers me.

since we work on different platforms, I will keep working with you on improving the test framework 🙂

Reply:
Ok, that's fine.

Reply:
why? the current code doesn't show helpful message when the terminal exits early. it only prints IO error and `<false> is not true.`

```
  "Input/output error @ io_fillbuf - fd:60 /dev/pts/25\n" +
  "[DEBUG SESSION LOG]\n" +
  "> [1, 1] in /tmp/debugger20210613-1699-38ryg3.rb\r\n" +
  "> =>    1| a = 1\r\n" +
  "> =>#0\t<main> at /tmp/debugger20210613-1699-38ryg3.rb:1\r\n" +
  "> \r\n" +
  "> (rdbg)\r\n" +
  "> continue\r\n" +
  ".\n" +
  "<false> is not true.">.
```

Reply:
Because, I think IO error and `[DEBUG SESSION LOG]` were really helpful while developing test framework. In addition, test will stop when the error occurs.
Do you think they're useless?

Reply:
imagine this: it's your first time contributing to this project and you know nothing about the test framework. 
by accident, you wrote an extra `continue` in the commands. after the test failed because of early exit, can you understand why based on the above message?

- what does `<false> is not true.` mean?
- why is there an `Input/output error @ io_fillbuf`?

my point is, to understand the current message, you need to understand how the test framework works. you and I understand it, but not everyone.

I'm not saying the original message is useless. it must be helpful in some context, but not in general because they're bound with the implementation detail.

since the most common scenario to trigger this error is because of early exit, why don't we just show that to reduce the confusion? 

`expect all commands/assertions to be executed. still have 3 left.`

this also isn't the final form of the test framework. for example, it'd be great to print information like:
- what's the last command before the termination
- what's the next command

so if you still find the original information useful in some context, please can post some logs to help me change the code. but don't just ask me to revert to the original one.

Reply:
Thank you for explaining to me. That makes sense. So, how about including the error information in `expect all commands/assertions to be executed. still have 3 left.`?

When I write test for remote debugger, this error occurs.
```shell
$ ruby test/debug/delete_test.rb
Loaded suite test/debug/delete_test
Started
F
================================================================================================================================================================================================================================================================================
Failure: test_deleting_a_breakpoint_removes_it_from_breakpoints_list(DEBUGGER__::DeleteTest):
  Input/output error @ io_writev - /dev/ttys000
  [DEBUG SESSION LOG]
  .
  <false> is not true.
/Users/[user_name]/workspace/debug/test/support/utils.rb:90:in `rescue in block in create_pseudo_terminal'
/Users/[user_name]/workspace/debug/test/support/utils.rb:55:in `block in create_pseudo_terminal'
/Users/[user_name]/workspace/debug/test/support/utils.rb:52:in `spawn'
/Users/[user_name]/workspace/debug/test/support/utils.rb:52:in `create_pseudo_terminal'
/Users/[user_name]/workspace/debug/test/support/utils.rb:33:in `debug_code'
test/debug/delete_test.rb:38:in `test_deleting_a_breakpoint_removes_it_from_breakpoints_list'
     35:     end
     36:
     37:     def test_deleting_a_breakpoint_removes_it_from_breakpoints_list
  => 38:       debug_code(program) do
     39:         type 'break if n == 1'
     40:         type 'continue'
     41:         # assert_line_num 8
================================================================================================================================================================================================================================================================================
```

I could find there was a problem in writable IO from this error.

However, if we change this code, it looks like
```shell
================================================================================================================================================================================================================================================================================
Failure: test_deleting_a_breakpoint_removes_it_from_breakpoints_list(DEBUGGER__::DeleteTest):
  expect all commands/assertions to be executed. still have 4 left.
  <#<Thread::Queue:0x00007f9f4d892990>> was expected to be empty.
/Users/[user_name]/workspace/debug/test/support/utils.rb:96:in `assert_empty_queue'
/Users/[user_name]/workspace/debug/test/support/utils.rb:86:in `rescue in block in create_pseudo_terminal'
/Users/[user_name]/workspace/debug/test/support/utils.rb:55:in `block in create_pseudo_terminal'
/Users/[user_name]/workspace/debug/test/support/utils.rb:52:in `spawn'
/Users/[user_name]/workspace/debug/test/support/utils.rb:52:in `create_pseudo_terminal'
/Users/[user_name]/workspace/debug/test/support/utils.rb:33:in `debug_code'
test/debug/delete_test.rb:38:in `test_deleting_a_breakpoint_removes_it_from_breakpoints_list'
     35:     end
     36:
     37:     def test_deleting_a_breakpoint_removes_it_from_breakpoints_list
  => 38:       debug_code(program) do
     39:         type 'break if n == 1'
     40:         type 'continue'
     41:         # assert_line_num 8
================================================================================================================================================================================================================================================================================
```

Above error makes it a little harder to understand cause of the error because there is no information about the error.

Reply:
@ono-max thank you for providing the detailed example, this is what I was asking for 🙂 
I've pushed an update to include exception's info (optional), can you give it a try?

Reply:
Thanks! This is great 👍 
```shell
Failure: test_next_steps_over_rescue_when_raising_from_method(DEBUGGER__::RescueControlFlowTest):
  expect all commands/assertions to be executed. still have 7 left.
  associated exception: Errno::EIO - Input/output error @ io_writev - /dev/ttys004.
  <#<Thread::Queue:0x00007ff42e01c9d8>> was expected to be empty.
/Users/naotto/workspace/debug/test/support/utils.rb:113:in `assert_empty_queue'
/Users/naotto/workspace/debug/test/support/utils.rb:99:in `rescue in block in create_pseudo_terminal'
/Users/naotto/workspace/debug/test/support/utils.rb:58:in `block in create_pseudo_terminal'
/Users/naotto/workspace/debug/test/support/utils.rb:55:in `spawn'
/Users/naotto/workspace/debug/test/support/utils.rb:55:in `create_pseudo_terminal'
/Users/naotto/workspace/debug/test/support/utils.rb:33:in `debug_code'
/Users/naotto/workspace/debug/test/debug/control_flow_commands_test.rb:211:in `test_next_steps_over_rescue_when_raising_from_method'
     208:     end
     209:
     210:     def test_next_steps_over_rescue_when_raising_from_method
  => 211:       debug_code(program) do
     212:         type 'break Foo::Bar.raise_error'
     213:         type 'continue'
     214:         assert_line_num 4
```